[13.0][MIG] l10n_es_dua_sii: Migration to 13.0

joao-p-marques opened the pull request:
@Tecnativa
TT26766

ping @Tardo @CarlosRoca13 @pedrobaeza

Jortolsa-S73 committed:
[ADD] l10n_es_dua_sii

omar7r committed:
[UPD] l10n_es_dua_sii: Adaptamos el módulo a la última versión del l10n_es_aeat_sii

Roodin committed:
[MIG] l10n_es_dua_sii: Migration to 10.0

mvrodriguez committed:
Translated using Weblate (Galician)

Currently translated at 100.0% (2 of 2 strings)

Translation: l10n-spain-10.0/l10n-spain-10.0-l10n_es_dua_sii
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-10-0/l10n-spain-10-0-l10n_es_dua_sii/gl/

angelmoya committed:
MIG l10n_es_dua_sii

Tardo committed:
[MIG] l10n_es_dua_sii: Migration to 12.0

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_sii: Fix logic + improve tests + contributors

eantones committed:
[FIX] l10n_es_dua_sii: SII tab hidden when DUA's fiscal position has not been defined

The "Fiscal position" field of the invoice is not a mandatory field so it might be not defined, in this case, its default value is "Régimen nacional".
The original module assumes that the fiscal position of the DUA always exists and this may not be so, mainly in multicompany environments where not all companies use DUA and they don't need to have this fiscal position created.

pedrobaeza committed:
[IMP] l10n_es_dua_sii: Don't depend on tax description

* README update. Not related but rechecking the roadmap was not accurated
* Tests adapted
* Tax description renamed

RodrigoBM committed:
[IMP] l10n_es_dua_sii: Remove <ImporteTotal> in DUA invoices

Se quita ImporteTotal como informacion que se manda al SII, para evitar
problemas de nuevas validaciones incorporadas en el SII para 2021

joao-p-marques committed:
[IMP] l10n_es_dua_sii: black, isort, prettier

pedrobaeza commented on l10n_es_dua_sii/models/account_move.py:
```diff
@@ -43,12 +44,12 @@ def _get_dua_fiscal_position_id(self, company):
             taxes = invoice._get_sii_taxes_map(["DUA"])
-            invoice.sii_dua_invoice = invoice.tax_line_ids.filtered(
-                lambda x: x.tax_id in taxes
+            invoice.sii_dua_invoice = invoice.invoice_line_ids.filtered(
+                lambda x: x.tax_ids and (tax_id in taxes for tax_id in x.tax_ids)
```
Esto no hace lo mismo que el anterior código, y sería bueno hacer el cast a boolean por si acaso:
```suggestion
            invoice.sii_dua_invoice = bool(invoice.invoice_line_ids.filtered(
                lambda x: any([tax in taxes for tax in x.tax_ids])))
```

pedrobaeza reviewed (changes requested):
En esta versión he refactorizado los tests y mejorado mucho su cobertura de casos funcionales, y la base de test en heredable:

https://github.com/OCA/l10n-spain/blob/87b3e689e2c3778a5806112d482ef732c9d4abf4/l10n_es_aeat_sii_oca/tests/test_l10n_es_aeat_sii.py#L28

Sería conveniente aprovecharlo para así no tener que crear ningún objeto contable (el test te da ya una compañía con plan contable español), y además, hacer un test del contenido final de lo mandado tal como se hace aquí:

https://github.com/OCA/l10n-spain/blob/87b3e689e2c3778a5806112d482ef732c9d4abf4/l10n_es_aeat_sii_oca/tests/test_l10n_es_aeat_sii.py#L164-L264

joao-p-marques committed:
[MIG] l10n_es_dua_sii: Migration to 13.0

- Refactor and improve tests
- 10n_es_aeat_sii_oca: adapt helper method

  - Allow for defining the base module from which to import the expected
    output json to allow usage downstream.
  - Return created invoice

TT26766

Co-Authored-By: Pedro M. Baeza <pedro.baeza@tecnativa.com>

joao-p-marques commented:
@pedrobaeza he hecho los cambios sugeridos y adaptado los tests.
He tenido que hacer un cambio en los metodos de ayuda de `l10n_es_aeat_sii_oca` para poder reutilizarlo. Si no debe ir aquí, tendría que copiar codigo y repetirlo en este modulo. 
Tampoco sé si debería ir en un PR aparte o puede ir en este.

pedrobaeza commented:
Ya que refactorizas el test base, devuelve la factura y así no tienes que crear otra para el resto de checks, y hazlo todo en un solo archivo/clase. Tampoco entiendo por qué necesitas 8 archivos JSON nuevos. Solo hay que testear el caso de este módulo, no otros. Con uno debería bastar.

joao-p-marques commented:
@pedrobaeza ya he hecho los cambios. Creo que era a esto que te referías, no?
Al final he dejado 2 ficheros JSON (1 para cada test) ya que de antes ya existían esos 2 casos.

Tardo reviewed (approved):
LGTM :+1: Code review

pedrobaeza commented:
He dado las últimas puntadas para dejar todo bien atado. Joao, puedes verlo en https://github.com/OCA/l10n-spain/compare/eb28ceffb83a006b21d42470dd94237fe9951a00..86ff0f8ce82250618061b6660bc0d394439d0c7d

/ocabot merge nobump
